src/loki/search.cc

```diff
@@ -492,24 +413,14 @@ struct bin_handler_t {
     }
 
     // assume its reachable
-    if (!check) {
-      return max_reach_limit;
-    }
-
-    // if you cant get through the end node then its not reachable since you cant leave the edge
-    const auto* node = reader.GetEndNode(edge, tile);
-    if (!node || node_filter(node)) {
-      return 0;
-    }
-
-    // store an index into cardinalities so we can tell when search paths merge
-    // if the index changes then we know its been merged with another path
-    // any node can reach itself so each of them starts at a reach of 1
-    size_t reach_index = reaches.size();
-    reach_indices[edge->endnode()] = reach_index;
-    reaches.push_back(1);
-    depth_first(tile, node, reach_index);
-    return reaches.back();
+    if (!check)
+      return {max_reach_limit, max_reach_limit};
+
+    // notice we do both directions here because in the end we use this reach for all input locations
+    auto reach =
+        SimpleReach(edge, max_reach_limit, reader, kInbound | kOutbound, edge_filter, node_filter);
+    directed_reaches[edge] = reach;
```

Comment:
here we should do one more optimization. if the opposing edge is not filtered and its endnode (this edges begin node) is not filtered, then we can assign the same reachability to the opposing edge and not have to do the reach check.